### PR TITLE
KAFKA-7058 [Connect] Comparing schema default values using Objects#deepEquals()

### PR DESCRIPTION
**Pull request by gunnarmorling**
https://issues.apache.org/jira/browse/KAFKA-7058
* Summary of testing strategy: Added new unit test

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java
Comment (gunnarmorling) [inline connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java:294]: Note this falls back to the regular `equals()` for non-array types. One might consider to either call `equals()` or `deepEquals()` based on the schema's type, but it felt like a potential bug lurking if future types are going to be added:

    (((type == ARRAY || type == BYTES) && Objects.deepEquals(defaultValue, schema.defaultValue)) ||
        Objects.equals(defaultValue, schema.defaultValue)) && ...

Happy to change it if you think it's preferable.

**Comment by gunnarmorling** [comment]
@rhauch ran into this one while working on some things in Debezium :)

File: connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java
Comment (rhauch) [reply connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java:294]: My only reservation is that `Objects.deepEquals` checks quite a few conditions before it falls back to the regular `equals()`. However, I agree that simpler is probably better, and in the grand scheme of things (no pun intended) the better approach.

**Review by rhauch — APPROVED** [review]
LGTM.

**Comment by rhauch** [comment]
@ewencp would you mind reviewing and merging. We're past code freeze, and this isn't a blocker. Since this will backport nicely back to 0.10.1 (code was last changed nearly [3 years ago](https://github.com/apache/kafka/pull/163), so it may be easier to wait until after the release to merge and backport. Up to you.

**Comment by rhauch** [comment]
Tests on JDK 10 are clean, but test failure on JDK 8 is unrelated: `kafka.api.AdminClientIntegrationTest.testMinimumRequestTimeouts`.

**Review by ewencp — APPROVED** [review]
LGTM, and looks like this can go all the way back to 0.9.0. It's a pretty fundamental issue for some schemas that causes caching in at least the `Converters` we're aware of to be broken/inefficient, so I think its worth backporting and considering a blocker, even if it isn't technically a regression.

Thanks @gunnarmorling for catching this!

**Pull request closed without merging**
